File: sdk/src/main/java/com/kin/ecosystem/main/presenter/EcosystemPresenter.java
package com.kin.ecosystem.main.presenter;


import static com.kin.ecosystem.Kin.KEY_ECOSYSTEM_EXPERIENCE;
import static com.kin.ecosystem.main.ScreenId.MARKETPLACE;
import static com.kin.ecosystem.main.ScreenId.NONE;
import static com.kin.ecosystem.main.ScreenId.ORDER_HISTORY;
import static com.kin.ecosystem.main.Title.MARKETPLACE_TITLE;
import static com.kin.ecosystem.main.Title.ORDER_HISTORY_TITLE;

import android.os.Bundle;
import android.support.annotation.NonNull;
import com.kin.ecosystem.EcosystemExperience;
import com.kin.ecosystem.base.BasePresenter;
import com.kin.ecosystem.common.Observer;
import com.kin.ecosystem.common.model.Balance;
import com.kin.ecosystem.core.data.blockchain.BlockchainSource;
import com.kin.ecosystem.core.data.settings.SettingsDataSource;
import com.kin.ecosystem.main.INavigator;
import com.kin.ecosystem.main.ScreenId;
import com.kin.ecosystem.main.Title;
import com.kin.ecosystem.main.view.IEcosystemView;
import java.math.BigDecimal;

public class EcosystemPresenter extends BasePresenter<IEcosystemView> implements IEcosystemPresenter {

	private static final String KEY_SCREEN_ID = "screen_id";
	private static final String KEY_CONSUMED_INTENT_EXTRAS = "consumed_intent_extras";
	private @ScreenId
	int visibleScreen;
	private @EcosystemExperience
	int experience;
	private boolean isConsumedIntentExtras;
	private final SettingsDataSource settingsDataSource;
	private final BlockchainSource blockchainSource;
	private INavigator navigator;

	private Observer<Balance> balanceObserver;
	private Balance currentBalance;

	public EcosystemPresenter(@NonNull IEcosystemView view, @NonNull SettingsDataSource settingsDataSource,
		@NonNull final BlockchainSource blockchainSource,
		@NonNull INavigator navigator, Bundle savedInstanceState, Bundle extras) {
		this.view = view;
		this.settingsDataSource = settingsDataSource;
		this.blockchainSource = blockchainSource;
		this.navigator = navigator;
		this.currentBalance = blockchainSource.getBalance();

		// Must come before processIntentExtras, so we can define if the intent was consumed already.
		processSavedInstanceState(savedInstanceState);
		processIntentExtras(extras);

		this.view.attachPresenter(this);
	}

	private void processSavedInstanceState(Bundle savedInstanceState) {
		this.visibleScreen = getVisibleScreen(savedInstanceState);
		this.isConsumedIntentExtras = getIsConsumedIntentExtras(savedInstanceState);
	}

	private boolean getIsConsumedIntentExtras(Bundle savedInstanceState) {
		return savedInstanceState != null && savedInstanceState.getBoolean(KEY_CONSUMED_INTENT_EXTRAS);
	}

	private int getVisibleScreen(Bundle savedInstanceState) {
		return savedInstanceState != null ? savedInstanceState.getInt(KEY_SCREEN_ID, NONE) : NONE;
	}

	private void processIntentExtras(Bundle extras) {
		if (!isConsumedIntentExtras) {
			this.experience = getExperience(extras);
			this.isConsumedIntentExtras = true;
		}
	}

	private int getExperience(Bundle extras) {
		return extras != null ? extras.getInt(KEY_ECOSYSTEM_EXPERIENCE, EcosystemExperience.MARKETPLACE)
			: EcosystemExperience.MARKETPLACE;
	}

	@Override
	public void onAttach(IEcosystemView view) {
		super.onAttach(view);
		if (experience == EcosystemExperience.ORDER_HISTORY) {
			experience = EcosystemExperience.MARKETPLACE;
			launchOrderHistory();
		} else {
			navigateToVisibleScreen(visibleScreen);
		}
	}

	private void launchOrderHistory() {
		if (view != null) {
			navigator.navigateToOrderHistory(false);
		}
	}

	private void navigateToVisibleScreen(int visibleScreen) {
		if (view != null) {
			switch (visibleScreen) {
				case ORDER_HISTORY:
					if (navigator != null) {
						navigator.navigateToOrderHistory(false);
					}
					break;
				case MARKETPLACE:
				case NONE:
				default:
					if (navigator != null) {
						navigator.navigateToMarketplace(false);
					}
					break;

			}
		}
	}

	@Override
	public void onStart() {
		blockchainSource.reconnectBalanceConnection();
		updateMenuSettingsIcon();
	}

	@Override
	public void onStop() {
		removeBalanceObserver();
	}

	@Override
	public void onSaveInstanceState(Bundle outState) {
		outState.putInt(KEY_SCREEN_ID, visibleScreen);
		outState.putBoolean(KEY_CONSUMED_INTENT_EXTRAS, isConsumedIntentExtras);
	}

	@Override
	public void onDetach() {
		super.onDetach();
		removeBalanceObserver();
		navigator = null;
	}

	private void addBalanceObserver() {
		removeBalanceObserver();
		balanceObserver = new Observer<Balance>() {
			@Override
			public void onChanged(Balance value) {
				currentBalance = value;
				if (isGreaterThenZero(value)) {
					updateMenuSettingsIcon();
				}
			}
		};
		blockchainSource.addBalanceObserver(balanceObserver, false);
	}

	private boolean isGreaterThenZero(Balance value) {
		return value.getAmount().compareTo(BigDecimal.ZERO) == 1;
	}

	private void updateMenuSettingsIcon() {
		if (!settingsDataSource.isBackedUp()) {
			if (isGreaterThenZero(currentBalance)) {
				changeMenuTouchIndicator(true);
				removeBalanceObserver();
			} else {
				addBalanceObserver();
				changeMenuTouchIndicator(false);
			}
		} else {
			changeMenuTouchIndicator(false);
		}
	}

	private void removeBalanceObserver() {
		if (balanceObserver != null) {
			blockchainSource.removeBalanceObserver(balanceObserver, false);
			balanceObserver = null;
		}
	}

	@Override
	public void balanceItemClicked() {
		if (view != null && visibleScreen != ORDER_HISTORY && navigator != null) {
			navigator.navigateToOrderHistory(false);
		}
	}

	@Override
	public void backButtonPressed() {
		if (view != null) {
			view.navigateBack();
		}
	}

	@Override
	public void visibleScreen(@ScreenId final int id) {
		visibleScreen = id;
		@Title final int title;
		switch (id) {
			case ORDER_HISTORY:
				title = ORDER_HISTORY_TITLE;
				break;
			case MARKETPLACE:
			default:
				title = MARKETPLACE_TITLE;
				break;
		}

		if (view != null) {
			view.updateTitle(title);
		}
	}

	private void changeMenuTouchIndicator(final boolean isVisible) {
		if (view != null) {
			view.showMenuTouchIndicator(isVisible);
		}
	}

	@Override
	public void settingsMenuClicked() {
		if (navigator != null) {
			navigator.navigateToSettings();
		}
	}

	@Override
	public void onMenuInitialized() {
		updateMenuSettingsIcon();
	}
}

File: sdk/src/main/java/com/kin/ecosystem/main/presenter/EcosystemPresenter.kt
package com.kin.ecosystem.main.presenter

import android.os.Bundle
import com.kin.ecosystem.EcosystemExperience
import com.kin.ecosystem.Kin.KEY_ECOSYSTEM_EXPERIENCE
import com.kin.ecosystem.R
import com.kin.ecosystem.base.BasePresenter
import com.kin.ecosystem.base.customAnimation
import com.kin.ecosystem.core.bi.EventLogger
import com.kin.ecosystem.core.bi.events.PageCloseTapped
import com.kin.ecosystem.core.data.auth.AuthDataSource
import com.kin.ecosystem.core.data.blockchain.BlockchainSource
import com.kin.ecosystem.core.data.settings.SettingsDataSource
import com.kin.ecosystem.main.INavigator
import com.kin.ecosystem.main.ScreenId
import com.kin.ecosystem.main.view.IEcosystemView

class EcosystemPresenter(private val authDataSource: AuthDataSource,
                         private val settingsDataSource: SettingsDataSource,
                         private val blockchainSource: BlockchainSource,
                         private val eventLogger: EventLogger,
                         private val navigator: INavigator?, savedInstanceState: Bundle?, extras: Bundle) : BasePresenter<IEcosystemView>(), IEcosystemPresenter {

	private var visibleScreen: ScreenId = ScreenId.NONE
	@EcosystemExperience
	private var experience: Int = EcosystemExperience.NONE
	private var isConsumedIntentExtras: Boolean = false

	init {
		// Must come before processIntentExtras, so we can define if the intent was consumed already.
		processSavedInstanceState(savedInstanceState)
		processIntentExtras(extras)
	}

	private fun processSavedInstanceState(savedInstanceState: Bundle?) {
		this.visibleScreen = getVisibleScreen(savedInstanceState)
		this.isConsumedIntentExtras = getIsConsumedIntentExtras(savedInstanceState)
	}

	private fun getIsConsumedIntentExtras(savedInstanceState: Bundle?): Boolean {
		return savedInstanceState?.getBoolean(KEY_CONSUMED_INTENT_EXTRAS) ?: run { false }
	}

	private fun getVisibleScreen(savedInstanceState: Bundle?): ScreenId {
		return ScreenId.valueOf(savedInstanceState?.getString(KEY_SCREEN_ID, ScreenId.NONE.name) ?: ScreenId.NONE.name)
	}

	private fun processIntentExtras(extras: Bundle) {
		if (!isConsumedIntentExtras) {
			this.experience = getExperience(extras)
			this.visibleScreen = getVisibleScreen(extras)
			this.isConsumedIntentExtras = true
		}
	}

	private fun getExperience(extras: Bundle?): Int {
		return extras?.getInt(KEY_ECOSYSTEM_EXPERIENCE, EcosystemExperience.NONE) ?: EcosystemExperience.NONE
	}

	override fun onAttach(view: IEcosystemView) {
		super.onAttach(view)
		if (visibleScreen == ScreenId.NOT_ENOUGH_KIN) {
			// first to show should be without animation
			navigator?.showNotEnoughKin(false)
		} else {
			val kinUserId = authDataSource.ecosystemUserID
			if (!settingsDataSource.isSawOnboarding(kinUserId)) {
				navigateToVisibleScreen(ScreenId.ONBOARDING)
			} else {
				if (experience == EcosystemExperience.ORDER_HISTORY) {
					experience = EcosystemExperience.NONE
					launchOrderHistory()
				} else {
					navigateToVisibleScreen(visibleScreen)
				}
			}
		}
	}

	private fun launchOrderHistory() {
		view?.let {
			navigator?.navigateToOrderHistory(customAnimation {
				enter = 0
				exit = R.anim.kinecosystem_slide_out_right
			}, addToBackStack = false)
		}
	}

	private fun navigateToVisibleScreen(visibleScreen: ScreenId) {
		view?.let {
			when (visibleScreen) {
				ScreenId.NOT_ENOUGH_KIN -> navigator?.showNotEnoughKin()
				ScreenId.ONBOARDING -> navigator?.navigateToOnboarding()
				ScreenId.ORDER_HISTORY -> navigator?.navigateToOrderHistory(customAnimation {
					enter = R.anim.kinecosystem_slide_in_right
					exit = R.anim.kinecosystem_slide_out_left
					popEnter = R.anim.kinrecovery_slide_in_left
					popExit = R.anim.kinecosystem_slide_out_right
				}, addToBackStack = false)
				ScreenId.MARKETPLACE,
				ScreenId.NONE ->
					navigator?.navigateToMarketplace()
				else -> navigator?.navigateToMarketplace(customAnimation {
					enter = R.anim.kinecosystem_slide_in_right
					exit = R.anim.kinecosystem_slide_out_right
				})
			}
		}
	}

	override fun onStart() {
		blockchainSource.reconnectBalanceConnection()
	}

	override fun touchedOutside() {
		navigator?.close()
		sendPageCloseEvent(PageCloseTapped.ExitType.BACKGROUND_APP)
	}

	override fun onSaveInstanceState(outState: Bundle) {
		with(outState) {
			putString(KEY_SCREEN_ID, visibleScreen.name)
			putBoolean(KEY_CONSUMED_INTENT_EXTRAS, isConsumedIntentExtras)
		}
	}

	override fun backButtonPressed() {
		view?.navigateBack()
		sendPageCloseEvent(PageCloseTapped.ExitType.ANDROID_NAVIGATOR)
	}

	private fun sendPageCloseEvent(exitType: PageCloseTapped.ExitType) {
		var pageName: PageCloseTapped.PageName? = null
		when (visibleScreen) {
			ScreenId.MARKETPLACE -> pageName = PageCloseTapped.PageName.MAIN_PAGE
			ScreenId.ORDER_HISTORY -> pageName = PageCloseTapped.PageName.MY_KIN_PAGE
			ScreenId.SETTINGS -> pageName = PageCloseTapped.PageName.SETTINGS
			ScreenId.ONBOARDING -> pageName = PageCloseTapped.PageName.ONBOARDING
			ScreenId.NOT_ENOUGH_KIN -> pageName = PageCloseTapped.PageName.DIALOGS_NOT_ENOUGH_KIN
		}
		eventLogger.send(PageCloseTapped.create(exitType, pageName))
	}

	override fun visibleScreen(id: ScreenId) {
		visibleScreen = id
	}

	companion object {
		const val KEY_SCREEN_ID = "screen_id"
		private const val KEY_CONSUMED_INTENT_EXTRAS = "consumed_intent_extras"
	}
}
